### PR TITLE
Update to libgit2 6aa06b6

### PR DESCRIPTION
**Pull request by carlosmn**
Update to 6aa06b6 with the speed fixes for merge and tree parsing.

### CODE REVIEW

File: LibGit2Sharp/Core/GitMergeOpts.cs [hunk 1]
```diff
@@ -98,37 +98,50 @@ internal enum GitMergePreference
...
-    internal enum GitMergeFileFlags
+    internal enum GitMergeFileFlag
     {
         /// <summary>
         /// Defaults
         /// </summary>
         GIT_MERGE_FILE_DEFAULT = 0,
 
-        /// <summary>
+        /// <summ   ary>
```
Comment (whoisj) [inline LibGit2Sharp/Core/GitMergeOpts.cs]: looks like a space error

File: LibGit2Sharp/Core/GitMergeOpts.cs [hunk 2]
```diff
@@ -98,37 +98,50 @@ internal enum GitMergePreference
...
+        /// a recursive merge base (by merging the multiple merge bases),
+        /// instead simply use the first base.  This flag provides a similar
+        /// merge base to `git-merge-resolve`.
+        /// </summary>
+        GIT_MERGE_NO_RECURSIVE = (1 << 3),
     }
 
     [Flags]
-    internal enum GitMergeFileFlags
+    internal enum GitMergeFileFlag
```
Comment (whoisj) [inline LibGit2Sharp/Core/GitMergeOpts.cs]: Why drop the 's' from `GitMergeFileFlags`, won't this be a breaking change with no benefit?

**Comment by whoisj** [comment]
generally looks good - left minor comments.

File: LibGit2Sharp/Core/GitMergeOpts.cs [hunk 3]
```diff
@@ -35,7 +35,7 @@ internal struct GitMergeOpts
         /// <summary>
         /// File merging flags.
         /// </summary>
-        public GitMergeFileFlags FileFlags;
+        public GitMergeFileFlag FileFlags;
```
Comment (ethomson) [inline LibGit2Sharp/Core/GitMergeOpts.cs]: This struct needs a `public uint RenameThreshold` between the `SimilarityMetric` and the `MergeFileFavorFlags`.

File: LibGit2Sharp/Core/GitMergeOpts.cs (hunk 2 above)
Comment (nulltoken) [inline LibGit2Sharp/Core/GitMergeOpts.cs]: >won't this be a breaking change with no benefit?

I think this type isn't publicly exposed :wink:

**Comment by nulltoken** [comment]
Shouldn't `package.config` be updated to 1.0.114?

File: LibGit2Sharp/Core/GitMergeOpts.cs (hunk 3 above)
Comment (carlosmn) [inline LibGit2Sharp/Core/GitMergeOpts.cs]: Oops, I missed that in the diff.

**Comment by carlosmn** [comment]
> Shouldn't package.config be updated to 1.0.114?

It should, but hotel wifi is its own special circle of hell, so after nuget failed to upload the package a few times, I decided I'd do it after I get back home.

File: LibGit2Sharp/Core/GitMergeOpts.cs (hunk 2 above)
Comment (carlosmn) [inline LibGit2Sharp/Core/GitMergeOpts.cs]: It's to mach the libgit2 rename, which brings this type into compliance with the naming structure.

**Comment by carlosmn** [comment]
I've now also added a `ConflictInMergeException` to throw when we return `GIT_EMERGECONFLICT`. On the next release, we can move that over to `MergeConflictException` and deprecate this one.

File: LibGit2Sharp/MergeOptionsBase.cs [hunk 4]
```diff
@@ -23,6 +23,16 @@ protected MergeOptionsBase()
         public bool FindRenames { get; set; }
 
         /// <summary>
+        /// If set, fail on the first conflict. No merged index will be created.
+        /// </summary>
+        public bool FailOnConflict { get; set; }
```
Comment (nulltoken) [inline LibGit2Sharp/MergeOptionsBase.cs]: If this switch actually triggers a throw, maybe should we describe the thrown exception in the xml doc? and maybe rename this member as `ThrowsOnFirstConflict` ?

File: LibGit2Sharp/MergeOptionsBase.cs [hunk 5]
```diff
@@ -23,6 +23,16 @@ protected MergeOptionsBase()
...
+
+        /// <summary>
+        /// Do not write the REUC extension on the generated index
+        /// </summary>
+        public bool SkipReuc { get; set; }
```
Comment (nulltoken) [inline LibGit2Sharp/MergeOptionsBase.cs]: I'm not sure that many people are at ease with REUC (Currently the very first result Google returns for "git reuc" is the source code of reuc.c in libgit2. Could we try and enhance a tad the xml doc to quickly describe what's the impact of setting (or not) this switch? Similarly, in order to be a little more expressive, how about renaming it to a less terse `DoNotStoreResolvedConflictDataIntoTheIndexReucSection`... ok maybe something shorter then ;-)

File: LibGit2Sharp/MergeOptionsBase.cs (hunk 4 above)
Comment (carlosmn) [inline LibGit2Sharp/MergeOptionsBase.cs]: Done (explanation), but I'd rather keep the libgit2 name for the option. It gets harder to know what you're doing if they don't match.

File: LibGit2Sharp/MergeOptionsBase.cs (hunk 5 above)
Comment (carlosmn) [inline LibGit2Sharp/MergeOptionsBase.cs]: Nobody needs to be at ease with it, they can just ignore it. I've added some copy to use the full name and explain when you might want to disable it. It shouldn't need more than that to recognise that you're not in an environment which matches when this option is meant to be used.

Comment (nulltoken) [inline LibGit2Sharp/MergeOptionsBase.cs]: Cool doco. Thanks!

>I've added some copy to use the full name and explain when you might want to disable it

"disable"? But it's false by default, isn't it?

File: LibGit2Sharp/MergeOptionsBase.cs (hunk 4 above)
Comment (nulltoken) [inline LibGit2Sharp/MergeOptionsBase.cs]: In order to help to user, shouldn't we describe the type of exception that's going to be thrown?

>It gets harder to know what you're doing if they don't match.

I agree, from a troubleshooting perspective, and we've done this for internal types.

However, `Throws` means something very clear in .NET land. `Fail` doesn't, and I'd rather stick with the idioms of the language when we're exposing a public type.

File: LibGit2Sharp/MergeOptionsBase.cs (hunk 5 above)
Comment (carlosmn) [inline LibGit2Sharp/MergeOptionsBase.cs]: Disable writing the extension, as in set this option to true.

File: LibGit2Sharp/MergeOptionsBase.cs (hunk 4 above)
Comment (carlosmn) [inline LibGit2Sharp/MergeOptionsBase.cs]: The new text does show which exception we throw. But there's of course the issue of whether we should throw at all, or if this should just be part of the merge result we return.

**Comment by carlosmn** [comment]
I figured throwing when the user explicitly asked us to error out is Bad Form, so now I've made it so we return the normal error condition, just without any actual conflict listing.

**Comment by nulltoken** [comment]
> I figured throwing when the user explicitly asked us to error out is Bad Form, so now I've made it so we return the normal error condition, just without any actual conflict listing.

I'm not sold on this. Thats looks like a disguised scenario for "Can I merge without a conflict?". And returning an empty list of conflicts seems like very weird.

The `ObjectDatabase` already exposes a `CanMergeWithoutConflict()` method. How about moving the fail-fast optimization there and not publicly exposing this option?

**Comment by carlosmn** [comment]
This is indeed a feature intended for that scenario, and yes we should make the already-existing `CanMergeWithoutConflict()` method take advantage of this. But I don't think that this means we should stop consumers of the library from using it in their own code. We use the library because of the control it gives us, regardless of the scenarios for which we originally intended the code.

Also note that `CanMergeWithoutConflict()` is also rather opinionated itself in what we assume the user will want. It does not activate finding renames (and offers the user no way to do so) so some acceptable merges will not work. It all depends on what the user is planing to do beyond calling this one method.

Regardless of whether we want to expose this option to users or hide it behind a very specific method, throwing when we've asked for the short-circuit on error seems wrong. In `CanMergeWithoutConflict()` we'd be converting a catch block into a boolean, which reeks of code-smell.

There is yet another reason why I don't want to make everyone go through `CanMergeWithoutConflict()`. It simply doesn't do what we want to do in the scenarios for which the short-circuit was created. It answers the question of *whether* you could merge, but it doesn't tell you what the result of a successful merge is. So I can't use it for the GitHub/VSTS/MyVeryOwnPRInterface use-case, since in order to take advantage of the quick negatives, I'd have to perform successful merges twice (which I'm just gonna go ahead and claim is the 80% case) twice, which would make this feature completely useless.

**Comment by ethomson** [comment]
> which would make this feature completely useless.

Exactly.

**Pull request merged**